MB Refactoring

bouzaghrane opened the pull request:
This branch is created from the causal-effect-estimation branch.

@timothyb0912 @hassanobeid1994

Commit:
Suggestion for refactoring FindLongDataDist

Commit:
More refactoring suggestions for functions

Commit:
High level suggestions to refactor notebook 5.0

Commit:
Suggestions to refactor notebook 6.0

Very similar to suggestions for notebook 5.0

bouzaghrane commented:
Thinking about the suggestions I have, I see that they have to purposes:
* Make each function have one simple functionality
* Minimize duplication across functions and code in general

Commit:
Suggestions on refactoring for:

Conditional statements and method parameters.

Commit:
non saved work from previous commit.

Commit:
started refactoring in separate cells

The original code is not changed. It will not be changed until tests are written.

Commit:
refactored functions for getting

alternative specific variable distributions. The original functions won't be overwritten until tests are written.

Commit:
added refactored functions for:

Trip Specific Variables

Commit:
Refactored function for:

Simulating data from parameters dictionary

Commit:
Refactored functions for:

Simulating availability of alternatives

Commit:
Removed deprecated functions

Commit:
Started rewriting functions to be refactored

while not deleting original functions until tests are written.

Commit:
Rewriting refactored functions for:

Functionality for getting distributions of alternative specific variables.
Also, fixed parameters in a function in notebook 3.0

Commit:
Added rewritten function for simulating

data for nodes without parents.

Commit:
cleaned rewritten functions

Commit:
Added rewritten functionf for:

fitting regressions and simulating availability.

Commit:
correctly compute causal effects &

Add assertion at the end of the notebook

Commit:
Test function DistNodeNoParent

Commit:
Refactored and tested the DistNodeNoParent function

bouzaghrane commented:
I just completely refactored the DistNodeNoParent function and wrote a test for it at the bottom of the notebook. The test is based on `unittest` and testing modules from `numpy`. Once I put all of my functions in a `.py` file, I will make the transition from `unittest` to `pytest` which should be very straightforward.

Commit:
Added test for SimNodeNoParent

Commit:
Refactored SimNodeNoParent &

commented out old version.

Commit:
Add test for SimulateAvailability &

Started rewriting SimulateAvailability function

Commit:
Refactored SimulateAvailability

timothyb0912 commented on notebooks/working/3.0-mab-independent-variable-simulation.md:
```diff
@@ -34,487 +47,269 @@ import random
+    ind_spec_dict = defaultdict(dict)
+    for ind_var in ind_spec:
+        # generate array of values for individual specific variable
+        var_val = pd.Series([(data_long.loc[data_long[obs_id_col] == x][ind_var].unique()[0]) for x in data_long[obs_id_col].unique()])
```
Can you shorten the length of this line?
```suggestion

        def extract_indiv_specific_data(obs_id, ind_var_col):
            indiv_mask = data_long[obs_id_col] == obs_id
            indiv_data_value =\
                data_long.loc[indiv_mask, ind_var_col].iloc[0]
            return indiv_data_value

        var_val =\
            pd.Series([extract_indiv_specific_data(x, ind_var)
                       for x in data_long[obs_id_col].unique()])
```

And note, I'd put the definition of `extract_indiv_specific_data` outside the for-loop.

timothyb0912 replied:
Actually, couldn't you (perhaps even more simply), generate `var_val` with some pandas manipulation as follows:
```suggestion
        var_val =\
            (data_long[[obs_id_col, ind_var]]
                .drop_duplicates(obs_id_col, inplace=False)
                .loc[:, ind_var]
                .reset_index(drop=True)
            )
```

bouzaghrane replied:
You're right! Much faster as well!

timothyb0912 commented on notebooks/working/3.0-mab-independent-variable-simulation.md:
```diff
@@ -34,487 +47,269 @@ import random
+    trip_spec_dict = defaultdict(dict)
+    for trip_var in trip_spec:
+        # generate array of values for trip specific variable
+        trip_var_data = pd.Series([(data_long.loc[data_long[obs_id_col] == x][trip_var].unique()[0]) for x in data_long[obs_id_col].unique()])
```
Similar comment to as above. Can you shorten the line length to 79 characters per line (max)?

Also here's a suggestion for generating trip_var_data.
```suggestion
        trip_var_data =\
            (data_long[[obs_id_col, trip_var]]
                .drop_duplicates(obs_id_col, inplace=False)
                .loc[:, trip_var]
                .reset_index(drop=True)
            )
```

bouzaghrane replied:
Thanks!

timothyb0912 commented on notebooks/working/3.0-mab-independent-variable-simulation.md:
```diff
@@ -34,487 +47,269 @@ import random
+    # alternatives for each observation - This column will
+    # be needed when converting to long data -- this will
+    # be moved to a different column
+    fake_choice = [random.choice(np.nonzero(a == 1)[0]) + 1 for a in np.array(AV_matrix)]
```
I find this line hard to understand upon first reading it.

I would separate the various parts of this line to show what one desires with each action. For example,
```suggestion
    available_alt_indices = lambda av_row: np.nonzero(av_row == 1)[0]
    fake_choice = [random.choice(available_alt_indices(a)) + 1 for a in np.array(AV_matrix)]
```

bouzaghrane replied:
Very true.

timothyb0912 commented on notebooks/working/3.0-mab-independent-variable-simulation.md:
```diff
@@ -34,487 +47,269 @@ import random
+def getConstantDistAlt(alt_var, alt_data, alt_name_dic, alt):
+    constant_dict = defaultdict(dict)
+    # Add name of alternative to variable and store distriburion & parameters
+    constant_dict[alt_var+'_'+alt_name_dic[alt]]['distribution'] = 'constant'
+    constant_dict[alt_var+'_'+alt_name_dic[alt]]['parameters'] = alt_data[alt_var].unique()
```
To the best extent possible, try to not repeat yourself. For example,
```suggestion
    alt_specific_var_name = alt_var + '_' + alt_name_dic[alt]
    constant_dict[alt_specific_var_name]['distribution'] = 'constant'
    constant_dict[alt_specific_var_name]['parameters'] = alt_data[alt_var].unique()
```

timothyb0912 replied:
As a matter of fact, you end up repeating this same pattern across multiple functions (for example, in `getCategoricalDistAlt` and `getContinuousDistAlt`).

This suggests that you should separate this logic into its own function. One might do:
```suggestion
    alt_specific_var_name =\
        get_alt_specific_variable_name(alt_var, alt_name_dict, alt)
    constant_dict[alt_specific_var_name]['distribution'] = 'constant'
    constant_dict[alt_specific_var_name]['parameters'] = alt_data[alt_var].unique()
```
You can define the separated function as follows:
```
def get_alt_specific_variable_name(var_name, alt_name_map, alt_id):
    return var_name + '_' + alt_name_map[alt_id]
```

bouzaghrane replied:
Got it! Whenever a logic repeats, isolate it.

timothyb0912 commented on notebooks/working/3.0-mab-independent-variable-simulation.md:
```diff
@@ -34,487 +47,269 @@ import random
+    np_array_range = np.arange(alt_data[alt_var].max()+1)
+    array_bincount = np.bincount(alt_data[alt_var])
+    probs = array_bincount / len(alt_data[alt_var])
+    categ_dict[alt_var+'_'+alt_name_dic[alt]]['parameters'] = [np_array_range,
+                                                               probs]
```
```suggestion
    alt_specific_var_name =\
        get_alt_specific_variable_name(alt_var, alt_name_dict, alt)
    categ_dict[alt_specific_var_name]['parameters'] =\
        [np_array_range, probs]
```

bouzaghrane replied:
👍

timothyb0912 commented on notebooks/working/3.0-mab-independent-variable-simulation.md:
```diff
@@ -34,487 +47,269 @@ import random
+    # Get the best distribution and store in dictionary
+    BestDict = fitter_object.get_best()
+    # Add name of alternative to variable and store distriburion & parameters
+    cont_dict[alt_var+'_'+alt_name_dic[alt]]['distribution'] = list(BestDict.items())[0][0]
+    cont_dict[alt_var+'_'+alt_name_dic[alt]]['parameters'] = list(BestDict.items())[0][1]
```
Same comment as above re the logic for creating the alternative specific variable name.

bouzaghrane replied:
Yes sir!

timothyb0912 commented on notebooks/working/3.0-mab-independent-variable-simulation.md:
```diff
@@ -34,487 +47,269 @@ import random
+    return var_type == 'categorical'
+
+## Function to get distribution for unique values:
+def getConstantDist(ind_var, var_val):
```
I would increase the generality of this function as well as `getCategoricalDist` and `getContinuousDist`.
```suggestion
def getConstantDist(var_name, var_val):
```

bouzaghrane replied:
done!

timothyb0912 commented on notebooks/working/3.0-mab-independent-variable-simulation.md:
```diff
@@ -34,487 +47,269 @@ import random
+    constant_dict[alt_var+'_'+alt_name_dic[alt]]['parameters'] = alt_data[alt_var].unique()
+    return constant_dict
+
+def getCategoricalDistAlt(alt_var, alt_data, alt_name_dic, alt):
```
I would get rid of the following functions:
1. `getConstantDistAlt`
2. `getCategoricalDistAlt`
3. `getContinuousDistAlt`

They can be combined with their corresponding functions without the `Alt` if you abstract the generation of the alternative specific variable name into a separate function as suggested [above](https://github.com/hassanobeid1994/tr_b_causal_2020/pull/89/files?file-filters%5B%5D=.md#r419154494).

bouzaghrane replied:
Done!

timothyb0912 commented on notebooks/working/3.0-mab-independent-variable-simulation.md:
```diff
@@ -34,487 +47,269 @@ import random
+        # generate array of values for trip specific variable
+        trip_var_data = pd.Series([(data_long.loc[data_long[obs_id_col] == x][trip_var].unique()[0]) for x in data_long[obs_id_col].unique()])
+        # Get distribution if variable is categorical
+        var_type = var_types[trip_var]
```
I believe you are relying on a global variable here (which you shouldn't do).
`var_types` is not defined as one of the function inputs, and it is not generated between lines 199 and 205. Try adding `var_types` to your inputs!

bouzaghrane replied:
Good catch! totally missed that.

timothyb0912 commented on notebooks/working/3.0-mab-independent-variable-simulation.md:
```diff
@@ -34,487 +47,269 @@ import random
+            else:
+                ind_var_dict = getCategoricalDist(ind_var, var_val)
+        else:
+            ind_var_dict = getContinuousDist(ind_var, var_val, cont_dists)
```
I believe you are relying on a global variable here (which you shouldn't do).
`cont_dists` is not defined as one of the function inputs, and it is not generated between lines 100 and 114. Try adding `cont_dists` to your inputs!

bouzaghrane replied:
Great catch!

timothyb0912 commented on notebooks/working/3.0-mab-independent-variable-simulation.md:
```diff
@@ -34,487 +47,269 @@ import random
+# Code for Trip Specific Variables
+# Loop around the different available alternatives
+def tripSpecDist(data_long, obs_id_col, trip_spec, cont_dists):
+    # Trip Specific Variable (maybe combine with individual specific variables)
```
👍 Yes please combine the two functions.

If you truly did have another identifier for trip-id's, then (I assume?) they would follow long-format conventions of not repeating identifying information across columns. In other words, simply passing in a `trip_id_col` would not be sufficient to filter the dataframe to 1 row per decision-maker per trip. You would need the `obs_id_col` and the `trip_id_col`, and you would need to change the logic used to get the data that you want.

As such, this function doesn't do what you would want it to do (get a distribution of values across all unique trips). At the moment, it only duplicates the logic that should be used in `indSpecDist`.

bouzaghrane replied:
Agreed. In this case, is our `tour_id` variable the `trip_id_col` we want? The confusion is just around whether the `tour_id` can be the same for multiple trips on the same tour or not.

timothyb0912 replied:
Yeah, in general, `tour_id` can be the same for multiple trips on the same tour. However, if I remember correctly, this dataset has been filtered such that there is one tour for going to work/school and one tour returning from work/school. Some exploratory data analysis may answer the question here! Let me know if, when sampling to choice==1, you see individuals with multiple rows of the same `tour_id`. I think you won't due to the pre-processing that occurred.

timothyb0912 commented on notebooks/working/3.0-mab-independent-variable-simulation.md:
```diff
@@ -34,487 +47,269 @@ import random
+        trip_var_data = pd.Series([(data_long.loc[data_long[obs_id_col] == x][trip_var].unique()[0]) for x in data_long[obs_id_col].unique()])
+        # Get distribution if variable is categorical
+        var_type = var_types[trip_var]
+        if isCategorical(var_type):
+            if isUnique(trip_var_data):
+            # If only one category
+                trip_spec_var_dic = getConstantDist(trip_var, trip_var_data)
+            else:
+            # If more than one category
+                trip_spec_var_dic = getCategoricalDist(trip_var, trip_var_data)
+        else:
+            # If not categorical but just one unique value
+            if isUnique(trip_var_data):
+                trip_spec_var_dic = getConstantDist(trip_var, trip_var_data)
+            # If not categorical but just one unique value
+            else:
+                trip_spec_var_dic = getContinuousDist(trip_var, trip_var_data, cont_dists)
```
This logic exists in `indSpecDist`, `altSpecDist`, and `tripSpecDist`.

It should be separated into its own function `get_distribution_and_param_dict(var_name, var_data, cont_dists=ALL_SCIPY_CONT_DISTS)`, with `ALL_SCIPY_CONT_DISTS` defined as a global variable.

bouzaghrane replied:
Agreed, I just need to figure out how to combine all the kwargs from different functions given that we need different arguments in different functions.

timothyb0912 reviewed (changes requested):
Hey @bouzaghrane !
Here are my (initial) overall comments on Notebook 3 in your PR.

## (General) Suggestions
1. Follow PEP8 style guidelines unless there's a good reason not too.
   1. Use snake casing rather than camel-casing for your function names.
   2. Adhere to line-lengths of 79 characters per line (so people can easily read your code when splitting screens).
2. See whether you can combine `altSpecDist` and `indSpecDist` into a single function `get_distribution_dicts`. Skip the suggestion if it's harder than is worth the effort, but my strong feeling is that these two functions are more similar than they are different. In particular, you may be able to abstract to a single function, with a kwarg of `name_dict=None`. Then within the secondary for-loop of the general function, you can have something along the lines of `var_name = spec_key if name_dict is None else name_dict[spec_key]`.
3. After the final set of functions is nailed down, add docstrings to all of them.

timothyb0912 commented on notebooks/working/6.0-mab-estimate-causal-effect.md:
````diff
@@ -40,443 +66,628 @@ import viz
-# 1. Define functions to be used in notebook
 
-```python
+## Define the main function
````
The entire `DistNodeNoParent` function is way cleaner than before. Good job! 👍

timothyb0912 commented on notebooks/working/6.0-mab-estimate-causal-effect.md:
```diff
@@ -536,151 +747,71 @@ def DistNodeNoParent(data_long,
-            # If not categorical but just one unique value
-            else:
-                # Use the Fitter library to fit distributions
-                # to the data
-                fitter_object = Fitter(data=trip_var,
-                                       distributions=cont_dists,
-                                       timeout=60)
-                fitter_object.fit()
-                # Get the best distribution and store in dictionary
-                BestDict = fitter_object.get_best()
-                params_dict.setdefault(var, {})['distribution'] = list(BestDict.items())[0][0]
-                params_dict.setdefault(var, {})['parameters'] = list(BestDict.items())[0][1]
-
+    trip_spec_dic_params = tripSpecDist(data_long, obs_id_col, trip_spec, var_types, cont_dists)
+    params_dict.update(trip_spec_dic_params)
```
Similar to my comment from notebook 3, can you combine the various functions into a new function,
`get_distribution_dicts`, with a kwarg `name_dict=None`? Then within the second for-loop of one's function, you can have a line that says `var_name = spec_key if name_dict is None else name_dict[spec_key]`. I believe that's the main differences between `indSpecDist`, `altSpecDist`, and `tripSpecDist`.

bouzaghrane replied:
Yes!

bouzaghrane replied:
Actually, I think that only the logic for finding the distributions should be isolated instead of all the functions. Because the logic for getting the data is different.
In that sense, we can write a different helper that generates data with its own kwargs as well.

timothyb0912 commented on notebooks/working/6.0-mab-estimate-causal-effect.md:
```diff
@@ -40,443 +66,628 @@ import viz
+                trip_spec_var_dic = getConstantDist(trip_var, trip_var_data)
+            # If not categorical but just one unique value
+            else:
+                trip_spec_var_dic = getConstantDist(trip_var, trip_var_data, cont_dists)
```
Across all three functions, `indSpecDis`, `altSpecDist`, and `tripSpecDist`, the logic in the highlighted lines is the same. This logic should be refactored out into a separate function, perhaps `get_distribution_and_param_dict(var_name, var_values, cont_dists=ALL_SCIPY_CONT_DISTS)` where you define `ALL_SCIPY_CONT_DISTS` as a global constant.

timothyb0912 commented on notebooks/working/6.0-mab-estimate-causal-effect.md:
```diff
@@ -40,443 +66,628 @@ import viz
+    trip_spec_dict = {}
+    for trip_var in trip_spec:
+        # generate array of values for trip specific variable
+        trip_var_data = pd.Series([(data_long.loc[data_long[obs_id_col] == x][trip_var].unique()[0]) for x in data_long[obs_id_col].unique()])
```
Same suggestion as notebook 3:
```suggestion
        trip_var_data =\
            (data_long[[obs_id_col, trip_var]]
                .drop_duplicates(obs_id_col, inplace=False)
                .loc[:, trip_var]
                .reset_index(drop=True)
            )
```

timothyb0912 commented on notebooks/working/6.0-mab-estimate-causal-effect.md:
```diff
@@ -40,443 +66,628 @@ import viz
+    return var_type == 'empirical'
 
-#     mapping_matrix = sparse_row_to_col_matrix.dot(sparse_row_to_col_matrix.T)
+def getConstantDistAlt(alt_var, alt_data, alt_name_dic, alt):
```
As in notebook 3, I'd suggest getting rid of all functions `getConstantDistAlt`, `getEmpiricalDistAlt`, `getCategoricalDistAlt`, and `getContinuousDistAlt` in favor of their corresponding functions without the `Alt` suffix.

To do this, can you make all the corresponding functions simply take `(var_name, var_data)` as inputs? Then, for the alternative specific cases, you can use the function
```
def get_alt_specific_variable_name(base_var_name, alt_name_map, alt_id):
    return base_var_name + '_' + alt_name_map[alt_id]
```
on the lines before you generate the dictionary of distribution name and parameters.

timothyb0912 commented on notebooks/working/6.0-mab-estimate-causal-effect.md:
```diff
@@ -40,443 +66,628 @@ import viz
+## Functions to replace code within
+## DistNodeNoParent
+## Function for checking length
+def isUnique(var_values):
```
Can you camel case your function names? Also, feel free to let me know if there's a reason to not do so that I'm missing!

bouzaghrane replied:
No specific reason. I'll snake case them.

timothyb0912 commented on notebooks/working/6.0-mab-estimate-causal-effect.md:
```diff
@@ -40,443 +66,628 @@ import viz
+    ind_spec_dict = {}
+    for ind_var in ind_spec:
+        # generate array of values for individual specific variable
+        var_val = pd.Series([(data_long.loc[data_long[obs_id_col] == x][ind_var].unique()[0]) for x in data_long[obs_id_col].unique()])  
```
From notebook 3:
```suggestion
        var_val =\
            (data_long[[obs_id_col, ind_var]]
                .drop_duplicates(obs_id_col, inplace=False)
                .loc[:, ind_var]
                .reset_index(drop=True)
            )
```

timothyb0912 commented on notebooks/working/6.0-mab-estimate-causal-effect.md:
```diff
@@ -704,43 +835,192 @@ def SimNodeNoParent(params_dict, size=1000):
+    # alternatives for each observation - This column will
+    # be needed when converting to long data -- this will
+    # be moved to a different column
+    fake_choice = [random.choice(np.nonzero(a == 1)[0]) + 1 for a in np.array(AV_matrix)]
```
To make this function easier to understand, I suggest the following.
It'll make it clear what `np.nonzero(a == 1)[0]` is doing.
```suggestion
    available_alt_indices = lambda av_row: np.nonzero(av_row == 1)[0]
    fake_choice = [random.choice(available_alt_indices(a)) + 1 for a in np.array(AV_matrix)]
```

timothyb0912 reviewed (changes requested):
This is a review of the portion of notebook 6 before the simulation.

The comments are mainly the same as my review of notebook 3.

Overall, I greatly agree with your general plan of what else needs to be refactored.
I would add an effort to at times combine functionality when possible. It's probably better to have a function that is trivially generalized by one or two kwargs, if it allows one to save duplicating large parts of two functions.

Commit:
rewritten functions:

find_long_dist_data &  sim_df

Commit:
refactored `simulate_availability` &

snake cased all functions.

Commit:
refactoring as follows:

- updating some variable names and fixing logic of assigning variable name (alt specific vs ind specific) in notebook 3.
- refactor the get_dist_node_no_parent function in notebook 6.

Commit:
change dictionary to defaultdict

Commit:
Refactored simulate_availability in notebook 6.0 &

modified the documentation for the same function in notebook 3.0

Commit:
Refactored sim_node_no_parent

& similar function from notebook 3.0

Commit:
Continued refactoring of simulation from dist fct

in both notebook 6.0 and notebook 3.0.

Commit:
Refactor the regression fitting functions

Commit:
Added documentation for:

regression fitting functions.

Commit:
Ordered parameters for getting variable dists &

Added documentation for function to get the alt specific variable name, and updated the test suite with updated function names.

Commit:
Refactored regression prediction code &

Created helper function to get predictions from fitted regressions

Commit:
Fixed old availability matrix sim code &

Fixed to create fake choice column

Commit:
Refactored regressions with perturbation

Commit:
Cleaned fitted regressions

Commit:
Refactoring helper for loop for simulation.

Commit:
Updated lin_reg_pred function to

account for an additional scale parameter for amplifying the causal effect.
Added test for regression fitting.

Commit:
Added test for the regression prediction function

@timothyb0912 please look at the test and let me know what you think. I tried to have the simplest test/test cases possible

Commit:
Adding parameter cells

Commit:
Isolate more fixed parameters in list

Commit:
Renaming of variables storing probabilities

Commit:
Code linting

Commit:
create .py files for functions used in notebook &

remove unused functions in notebook.

Commit:
Added test suite to folder structure.

timothyb0912 commented:
@bouzaghrane Am I correct in thinking this branch can be closed?

bouzaghrane commented:
@timothyb0912 yes. Just closed it.
